spec/formatting.md

```diff
@@ -18,3 +18,38 @@ the local variable takes precedence.
 
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
+
+## Bidirectional Isolation
+
+Formatted messages may be bidirectional,
+i.e. contain a mix of left-to-right and right-to-left content.
+In all output formats,
+such parts should be isolated from each other so that
+the directionality of one part of the message does not affect
+the presentation of the rest of the message.
+
+Implementations may provide one or more strategies for isolation,
+but one such strategy must behave as follows,
+when formatting the message as a string:
+
+Considering a "formatted part" to be the formatted string representation of a Placeholder,
+when the formatted part has a direction different from the message as a whole
+or when its directionality is not known,
+the formatted string representation of the message must
+prefix the formatted part with one of the following isolate characters:
```

Comment:
This leaves a gap for same-base-paragraph-direction placeholders with various edge effects (generally starting or ending with weak runs or runs with neighboring character issues). I admit that we've wrapped ourselves up trying to define when we are required to isolate without having to run the bidi algorithm. I don't have a solution yet, so perhaps keeps this for now, but add a note about the issue??

I would split the term definition of "formatted part" from the requirement.

Reply:
I don't think I see the hole you're describing. Could you give an example?

Reply:
The pattern string is:

> السعر {$price} + {$shipping_amt} الشحن

The currency code for both is `USD` and the symbol for that is `USD` in the `ar-AE` locale. The resulting string is:

> 
السعر 1,234.56 AED + 12.99 USD الشحن

It should be:

> السعر ⁧1,234.56 USD⁩ + ⁧12.99 USD⁩ الشحن

Here are the code point sequences with and without RLI/PDI around the prices:

```
without: \u0627\u0644\u0633\u0639\u0631 1,234.56 USD + 12.99 USD \u0627\u0644\u0634\u062d\u0646
with: \u0627\u0644\u0633\u0639\u0631 \u20671,234.56 USD\u2069 + \u206712.99 USD\u2069 \u0627\u0644\u0634\u062d\u0646
```

In both cases the base paragraph direction of the pattern string *and also* each of the placeables is RTL, but the appearance of weak (numbers) and strong (letters) LTR runs confuses bidi unless isolated.

Dates often have this sort of thing too or any string that starts or ends with numeric sequences or enclosing punctuation.

Reply:
Ok, I see it now, how the formatted value might be nominally RTL but only composed of neutral and LTR characters. It's interesting to play around with this in different browsers, as I see this:

```js
const usd = new Intl.NumberFormat('ar-AE', { style: 'currency', currency: 'USD'}).format(12.99)

// in Chrome & Node.js
usd === "US$ 12.99"
usd.split('').map(c=>c.charCodeAt(0))
// [ 85, 83, 36, 160, 49, 50, 46, 57,  57 ]

// in Firefox
usd === "‏12.99 US$"
usd.split('').map(c=>c.charCodeAt(0))
// [ 8207, 49, 50, 46, 57, 57, 160, 85, 83, 36 ]
```

Specifically, Firefox prefixes the string with an RTL mark for currencies. It doesn't really help with the isolation, though.

So I agree that this situation is left outside the "must" statement, I think it's potentially ok? As in, at least we don't have a "must not" so that an implementation is allowed to introspect the formatted part and determine that it ought to be isolated.